Add `aarch64-pc-windows-gnullvm` and `x86_64-pc-windows-gnullvm` targets

mati865 opened the pull request:
Creating new target as told to do so in https://github.com/microsoft/windows-rs/pull/1846#pullrequestreview-1026401722

Fixes #1881

mati865 commented:
I don't know what is wrong but git keeps interpreting `crates/targets/aarch64_gnullvm/lib/libwindows.a` as a text file.

kennykerr commented:
> I don't know what is wrong but git keeps interpreting `crates/targets/aarch64_gnullvm/lib/libwindows.a` as a text file.

That's odd, the `.gitattributes` includes `*.a` so it shouldn't. 

https://github.com/microsoft/windows-rs/blob/4ba369acb36218429e3ea38c8a1660c47677a323/.gitattributes#L7

kennykerr commented:
<img width="616" alt="image" src="https://user-images.githubusercontent.com/9845234/178151762-60907fec-42c6-4b34-af50-a5f079f18c35.png">

It is suspiciously small.

mati865 commented:
> <img alt="image" width="616" src="https://user-images.githubusercontent.com/9845234/178151762-60907fec-42c6-4b34-af50-a5f079f18c35.png">
> 
> It is suspiciously small.

Yeah, noticed it as well before opening the PR but I had compared outputs of `llvm-nm crates/targets/aarch64_gnullvm/lib/libwindows.a` and `llvm-nm crates/targets/aarch64_msvc/lib/windows.lib` and they were mostly the same.
At least when it comes to symbols count they are the same:
```
$ llvm-readobj --coff-exports crates/targets/aarch64_gnullvm/lib/libwindows.a | grep Symbol | wc -l
34210

$ llvm-readobj --coff-exports crates/targets/aarch64_msvc/lib/windows.lib | grep Symbol | wc -l
34210
```

riverar commented on crates/tools/gnullvm/readme.md:
````diff
@@ -0,0 +1,16 @@
+The Windows umbrella lib (targeting mingw-w64 LLVM tooling) is generated using the following steps:
+
+0. Ensure MSYS2 MinGW environment is installed (https://www.mingw-w64.org/downloads/)
+1. Ensure `~\.cargo\config` has the following block to override toolchain defaults (customize paths as needed):
+
+    ```toml
+    [target.x86_64-pc-windows-gnu]
+    linker = "C:\\msys64\\mingw64\\bin\\x86_64-w64-mingw32-gcc.exe"
+    ar = "C:\\msys64\\mingw64\\bin\\ar.exe"
+    ```
+
+2. Open `MSYS2 MinGW 64-bit`
+3. Execute: `pacman -Syuu --noconfirm` (repeat until no further updates available)
+4. Execute `pacman --needed -S mingw-w64-x86_64-llvm`
+5. Navigate to crate root
+6. Execute: `PATH=$USERPROFILE/.cargo/bin:$PATH cargo run -p tool_gnullvm --target x86_64-pc-windows-gnu -- all`
````
Can we use clang directly? Would like to avoid piggy backing off the gnu target/gcc as much as possible. Any general pointers/tips on building your toolchain locally for testing?

mati865 replied:
> Can we use clang directly?

There would be no problem in using Clang as the linker for `x86_64-pc-windows-gnu` target but it would still use all other GNU tools like `ld.bfd`, `libgcc`, etc.

To be honest it doesn't matter which toolchain we use here, it can be any target since we use it only to build `tool_gnullvm` so I based it off `tool_gnu`.
Tools called inside `tool_gnullvm` are unaffected by target that built it (with some changes one could even run it using Linux target).

riverar replied:
That's OK, just trying to use as little of that _older stuff_ as possible. 😄

I was able to run the tool with no problems with your new target + rustc nightly, so I'd go for that!

riverar commented:
Any general pointers/tips on building your toolchain locally for testing? I'm going through the steps of compiling `rustc` now, which I assume is the starting point.

mati865 commented:
You could try first if **unstable** `-Z build-std` feature will suffice your needs, it's the easiest way to play with this target right now.
For an example you can even run `tool_gnullvm` using `x86_64-pc-windows-gnullvm` target with these steps:
1. Open MSYS2 CLANG64 environment
2. Run `pacman --needed -S mingw-w64-clang-x86_64-toolchain`
3. Run `PATH=$(cygpath -u $USERPROFILE)/.cargo/bin:$PATH cargo r --target x86_64-pc-windows-gnullvm -Z build-std -p tool_gnullvm -- all`
It requires nightly Rust though.

If you want full toolchain I'll provide you the exact steps on how to build it today/tomorrow evening CEST.

mati865 commented on crates/tools/gnullvm/src/main.rs:
```diff
@@ -0,0 +1,125 @@
+    for (function, params) in functions {
+        if dlltool_target.eq("i386") {
+            def.write_all(format!("{}@{}\n", function, params).as_bytes()).unwrap();
+        } else {
```
Sorry, missed it.

riverar commented:
Ah cool, I didn't realize this made it into nightly. 👍 No full toolchain steps needed for our purposes at this time, although it might be handy [to add to the docs](https://doc.rust-lang.org/nightly/rustc/platform-support/pc-windows-gnullvm.html) at some point.

Compiling `tool_gnullvm` (targeting `*-gnullvm`) results in libunwind failures. I see clang was invoked with `--unwindlib=none` so not sure what's going on there yet.

<details>
<summary>Full output:</summary>

```
$ PATH=$(cygpath -u $USERPROFILE)/.cargo/bin:$PATH cargo +nightly r --target x86_64-pc-windows-gnullvm -Z build-std -p tool_gnullvm -- all
   Compiling tool_gnullvm v0.0.0 (C:\Sources\windows-rs\crates\tools\gnullvm)
error: linking with `C:\msys64\mingw64\bin\clang.exe` failed: exit code: 1
  |
  = note: "C:\\msys64\\mingw64\\bin\\clang.exe" "-nolibc" "--unwindlib=none" "-m64" "C:\\Users\\Rafael\\AppData\\Local\\Temp\\rustcC65uNT\\symbols.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.12ezno69lpmyuu43.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.182fhvj1u3z7iewp.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.18libnrriw98oxa5.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.19coswx6llvdnn7o.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.19u62jculx1rys4c.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.19zyo4y10v97pxyc.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.1c7w2c0kmw4vax41.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.1fhaabranfwg26qr.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.1kb989b60cwsp40k.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.1lmdnorj2y13sjqj.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.1lufaybp8bgbxryf.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.1rscad6yb11zdupb.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.1v0oocqexrv9smmf.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.1zmgck2so313w7wv.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.23fsy0jzoccnj3gw.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.25so6al963q2emyd.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.26bn02cwg0enjfjp.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.2awck9hsuryladm7.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.2carsqilgwqywkd8.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.2e7djchnlh0ww7o7.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.2hhaykeo7gz5yd5y.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.2s7xba4l2cafbheb.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.2utj8qikjdawum8r.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.2wotsgfes5l9gc8u.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.2xb3pn7who3lfhvs.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.31tcqypfx8sy7tho.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.34jalpnpzia2w46x.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3c62t39ncatq5vlz.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3chyhombgrxzrnat.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3dsup3ekjghugufa.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3iju24uwbpdjkow4.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3jgisacdr7dgejlt.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3kdd597cu0ixmadb.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3lumt0z74ue70uuy.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3nnqtye7nfbh53f7.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3obygl2afdjpl1s6.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3s6g8dpesmhs2c6g.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3sbotzh55gfflxdt.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3sjwyk8dpeo5642h.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.3zf7avs5a1a4m75u.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.44yyo97bgsjpxa5x.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.46b1ysswldryyxb.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.4e3f01fox28igrwd.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.4e9imsxiyinfp08x.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.4fdvsv6s7yjfgy6g.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.4itxrucu3etl472e.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.4j7jlutq26n8b89l.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.4jryu1iwa0j2t8r9.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.4mvkjw2fjzfehgr8.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.4n8a5elnbwh0ktbb.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.4xkn8ckyjulx8bds.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.523rtzupxsdg8d54.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.56hrao22h6eux2f7.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.5b6w1bx5ttxo18kz.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.5ecxczoc7p2fv2qq.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.9oe63endkav5e41.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.ay7j6ko52ije6fb.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.iy05wnmkx473440.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.wkvi2cqd3pce0sv.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.z34xp5dqywgslxi.rcgu.o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.ge5v7etyc1jxeve.rcgu.o" "-L" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps" "-L" "C:\\Sources\\windows-rs\\target\\debug\\deps" "-L" "C:\\Users\\Rafael\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-gnullvm\\lib" "-Wl,-Bstatic" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libwindows_metadata-c91907ac2b553472.rlib" "-Wl,--start-group" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libstd-e9f2efaad10a3b98.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libpanic_unwind-5a5bf608ddb8ed39.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libobject-e23b7a7cfbdf5916.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libmemchr-96c6a8deb0ac1412.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libaddr2line-1667495d10f7223f.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libgimli-e31847701079c437.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\librustc_demangle-00c752d0d065974d.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libstd_detect-ad54cec87987c4a2.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libhashbrown-56e9dd1d6fe90939.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libminiz_oxide-2633874641d74771.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libadler-0e1435af86fb862f.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\librustc_std_workspace_alloc-dcaf45f9a093f678.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libunwind-c43da1cf3a899fc0.rlib" "-lunwind" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libcfg_if-e4d4a9cbb956fd9a.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\liblibc-69d6255466bb7ddf.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\liballoc-c5571d2b88345384.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\librustc_std_workspace_core-ad0baf2b5b0e44ce.rlib" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libcore-6a89cba04688d323.rlib" "-Wl,--end-group" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\libcompiler_builtins-94ce661bcd27ab31.rlib" "-Wl,-Bdynamic" "-ladvapi32" "-luserenv" "-lkernel32" "-lws2_32" "-lbcrypt" "-lmingw32" "-lmingwex" "-lmsvcrt" "-lkernel32" "-luser32" "-Wl,--nxcompat" "-L" "C:\\Users\\Rafael\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-gnullvm\\lib" "-o" "C:\\Sources\\windows-rs\\target\\x86_64-pc-windows-gnullvm\\debug\\deps\\tool_gnullvm-e1caceb089630d49.exe" "-Wl,--gc-sections" "-no-pie"
  = note: clang: warning: argument unused during compilation: '-nolibc' [-Wunused-command-line-argument]
          clang: warning: argument unused during compilation: '--unwindlib=none' [-Wunused-command-line-argument]
          clang: warning: argument unused during compilation: '-no-pie' [-Wunused-command-line-argument]
          C:/msys64/mingw64/bin/ld: cannot find -lunwind: No such file or directory
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

</details>

kennykerr commented on crates/tools/gnullvm/src/main.rs:
```diff
@@ -0,0 +1,121 @@
+        return;
+    };
+
+    let files = vec![metadata::reader::File::new("crates/libs/metadata/default/Windows.winmd").unwrap(), metadata::reader::File::new("crates/libs/metadata/default/Windows.Win32.winmd").unwrap(), metadata::reader::File::new("crates/libs/metadata/default/Windows.Win32.Interop.winmd").unwrap()];
+    let reader = &metadata::reader::Reader::new(&files);
+
+    let mut libraries = BTreeMap::<String, BTreeMap<String, usize>>::new();
+    let root = reader.tree("Windows.Win32", &[]).expect("`Windows` namespace not found");
+    for tree in root.flatten() {
+        if let Some(apis) = reader.get(metadata::reader::TypeName::new(tree.namespace, "Apis")).next() {
+            for method in reader.type_def_methods(apis) {
+                let impl_map = reader.method_def_impl_map(method).expect("ImplMap not found");
+                let scope = reader.impl_map_scope(impl_map);
+                let library = reader.module_ref_name(scope).to_lowercase();
+                let params = reader.method_def_size(method);
+                libraries.entry(library).or_default().insert(reader.method_def_name(method).to_string(), params);
+            }
+        }
+    }
```
All of this can now be replaced with `libraries()` from the `tool_lib` crate.

mati865 replied:
Copied this code from `tool_gnu`, should I also update it as well (in another PR)?

riverar replied:
@mati865 Is your choice 👍

kennykerr commented:
Is there no 32-bit target for gnullvm?

mati865 commented:
> Is there no 32-bit target for gnullvm?

ARM and i686 might be added in the future, right now I skipped them because Rust breaks their exception handling (the "glue" code is very specific to libgcc and doesn't work with LLVM's libunwind). 64-bit targets use SEH and are free of this issue.

mati865 commented:
> C:/msys64/mingw64/bin/ld

@riverar could you try using CLANG64 subsystem instead of MINGW64?

riverar commented:
@mati865 Yep, that works great. I'll add a review comment to the README then.

riverar commented on crates/tools/gnullvm/readme.md:
````diff
@@ -0,0 +1,16 @@
+The Windows umbrella lib (targeting mingw-w64 LLVM tooling) is generated using the following steps:
+
+0. Ensure MSYS2 MinGW environment is installed (https://www.mingw-w64.org/downloads/)
+1. Ensure `~\.cargo\config` has the following block to override toolchain defaults (customize paths as needed):
+
+    ```toml
+    [target.x86_64-pc-windows-gnu]
+    linker = "C:\\msys64\\mingw64\\bin\\x86_64-w64-mingw32-gcc.exe"
+    ar = "C:\\msys64\\mingw64\\bin\\ar.exe"
+    ```
````
This can be deleted.

riverar commented on crates/tools/gnullvm/readme.md:
````diff
@@ -0,0 +1,16 @@
+    ar = "C:\\msys64\\mingw64\\bin\\ar.exe"
+    ```
+
+2. Open `MSYS2 MinGW 64-bit`
````
```suggestion
2. Execute: `C:\msys64\clang64.exe`
```

riverar commented on crates/tools/gnullvm/readme.md:
```diff
@@ -0,0 +1,16 @@
+
+2. Open `MSYS2 MinGW 64-bit`
+3. Execute: `pacman -Syuu --noconfirm` (repeat until no further updates available)
+4. Execute `pacman --needed -S mingw-w64-x86_64-llvm`
```
Should probably add `mingw-w64-x86_64-clang`. do we need `mingw-w64-clang-x86_64-toolchain` as well?

```suggestion
4. Execute `pacman --needed -S mingw-w64-x86_64-llvm`
```

mati865 replied:
`mingw-w64-clang-x86_64-toolchain` is a bit overkill: https://packages.msys2.org/group/mingw-w64-clang-x86_64-toolchain as it installs everything that might be required for simple C/C++ projects (including Clang and libunwind).

If you want I can find out minimal required set of packages, probably it'd be: `pacman --needed -S mingw-w64-x86_64-clang mingw-w64-x86_64-libunwind`

mati865 commented on crates/tools/gnullvm/readme.md:
```diff
@@ -0,0 +1,8 @@
+2. Execute: `pacman -Syuu --noconfirm` (repeat until no further updates available)
+3. Execute `pacman --needed -S mingw-w64-clang-x86_64-toolchain`
+4. Navigate to crate root
+5. Execute: `PATH=$(cygpath -u $USERPROFILE)/.cargo/bin:$PATH cargo run -p tool_gnullvm --target x86_64-pc-windows-gnu -- all`
```
Tweaked this line to convert path with `cygpath` because MSYS2 doesn't understand Win32 style path from `$USERPROFILE`:
```
$ PATH=$USERPROFILE/.cargo/bin:$PATH

$ which cargo
which: no cargo in (C:\Users\mateusz/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/c/Windows/System32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0/:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl)

$ echo $PATH
C:\Users\mateusz/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/c/Windows/System32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0/:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl

$ PATH=$(cygpath -u $USERPROFILE)/.cargo/bin:$PATH

$ which cargo
/c/Users/mateusz/.cargo/bin/cargo

$ echo $PATH
/c/Users/mateusz/.cargo/bin:C:\Users\mateusz/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/c/Windows/System32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0/:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
```

riverar commented on .github/workflows/build.yml:
```diff
@@ -126,10 +126,12 @@ jobs:
           cargo clippy -p sample_spellchecker &&
           cargo clippy -p sample_uiautomation &&
           cargo clippy -p sample_xml &&
+          cargo clippy -p windows_aarch64_gnulllvm &&
           cargo clippy -p windows_aarch64_msvc &&
           cargo clippy -p windows_i686_gnu &&
           cargo clippy -p windows_i686_msvc &&
           cargo clippy -p windows_x86_64_gnu &&
+          cargo clippy -p windows_x86_64_gnulllvm &&
```
Typo: `windows_aarch64_gnulllvm` => `windows_aarch64_gnullvm`

kennykerr replied:
Note the build script is generated using `cargo run -p tool_yml`.

mati865 commented:
Ran `tool_yml` and updated readme with instructions on using `gnullvm` target on nightly.

I can squash commits manually and force-push once you are satisfied with this PR.

kennykerr commented:
> I can squash commits manually and force-push once you are satisfied with this PR.

Not to worry, we always "squash and merge" when completing.

riverar commented:
Looking pretty good!

@mati865 Can you update our rudimentary cross-compile CI tests here to test this new target? https://github.com/microsoft/windows-rs/blob/master/.github/workflows/cross-compile.yml

One possible [matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs) strategy could look like:
```yaml
matrix:
  image: [macos-latest, ubuntu-latest]
  version: [stable, nightly]
  target: [x86_64-pc-windows-gnu, xxxxxx_gnullvm]
  exclude:
  - target: xxxxxx_gnullvm
    version: stable
```

If you don't feel comfortable with that, need help, etc. just let us know.

mati865 commented:
@riverar linking the executable will fail without manually installing mingw-w64+LLVM toolchain like this one: https://github.com/mstorsjo/llvm-mingw
Should I do it?

riverar commented:
🎉 Passing the smoke tests, sweet!

I don't think this is specific to your gnullvm target but I'm wondering if the LLVM/mingw-w64 toolchain can be tweaked to leverage [Hybrid CRT](https://github.com/microsoft/WindowsAppSDK/blob/main/docs/Coding-Guidelines/HybridCRT.md) semantics to eliminate the dynamic CRT external dependency. I don't think this is a blocker for adoption but that's not my call.


Also, can you help me understand the differences between mstorsjo/llvm-mingw and me installing `mingw-w64-clang-...-toolchain` on Windows? Apologies in advance if that's obvious.

mati865 commented:
> I don't think this is specific to your gnullvm target but I'm wondering if the LLVM/mingw-w64 toolchain can be tweaked to leverage [Hybrid CRT](https://github.com/microsoft/WindowsAppSDK/blob/main/docs/Coding-Guidelines/HybridCRT.md?rgh-link-date=2022-07-13T00%3A37%3A50Z) semantics to eliminate the dynamic CRT external dependency.

I don't think so, mingw-w64 doesn't support linking static CRT at all because it doesn't have access to Microsoft's source codes nor static libraries. So I'd expect this Hybrid thing to have the same problem.

> Also, can you help me understand the differences between mstorsjo/llvm-mingw and me installing `mingw-w64-clang-...-toolchain` on Windows? Apologies in advance if that's obvious.

MSYS2 (the tool that provides `mingw-w64-clang-...-toolchain`) works only on Windows and provides package manager with toolchain, tools and libraries so it's similar to developing on Linux.
`llvm-mingw` is just a compressed toolchain so mingw-w64 + LLVM.

MSYS2's `CLANG*` subsystems are almost the same as `llvm-mingw` so both are fully compatible.
`MINGW*` and `UCRT*` subsystems however carry few patches making Clang/LLVM more compatible with GCC but incompatible with unpatched Clang/LLVM (like one in `CLANG*` subsystems or `llvm-mingw`).
This is especially visible when using TLS or exceptions inside executable.

So TL;DR `llvm-mingw` supports all major systems as the host but provides toolchain only, MSYS2 supports only Windows as the host but provides more complete Linux-like environment. You can use either of them and `cargo b --target x86_64-pc-windows-gnullvm -Z build-std` will work.

riverar commented on .github/workflows/cross-compile.yml:
```diff
@@ -45,10 +52,28 @@ jobs:
         run: brew install mingw-w64
         if: startsWith(matrix.image, 'macos-') && matrix.target == 'x86_64-pc-windows-gnu'
 
+      - name: Install LLVM MinGW toolchain
+        run: |
+          name='llvm-mingw-20220323-ucrt-ubuntu-18.04-x86_64'
+          curl -L -o $name.tar.xz https://github.com/mstorsjo/llvm-mingw/releases/download/20220323/$name.tar.xz
+          tar -xf $name.tar.xz
+          echo "$name/bin" >> $GITHUB_PATH
+        if: startsWith(matrix.image, 'ubuntu-') && contains(matrix.target, 'gnullvm') && matrix.version == 'nightly'
```
With the number of daily builds, we should be good Internet citizens and utilize a [cache](https://github.com/actions/cache) here.

Something like this should do (untested, probably not functional as-is):

```yaml
env:
  LLVM-MINGW-TOOLCHAIN-NAME: llvm-mingw-20220323-ucrt-ubuntu-18.04-x86_64
# ...
- name: LLVM MinGW toolchain cache configuration
  id: cache-llvm-mingw-toolchain
  uses: actions/cache@v3
  with:
    path: ${ llvm-mingw-toolchain-name } # Assumes .tar.xz extracted in working dir
    key: ${ llvm-mingw-toolchain-name }

- name: Install LLVM MinGW toolchain
  if: startsWith(matrix.image, 'ubuntu-') && contains(matrix.target, 'gnullvm') && matrix.version == 'nightly' && steps.cache-llvm-mingw-toolchain.outputs.cache-hit != 'true'
  run: |
    curl -L -o ${ llvm-mingw-toolchain-name }.tar.xz https://github.com/mstorsjo/llvm-mingw/releases/download/20220323/${ llvm-mingw-toolchain-name }.tar.xz
    tar -xf ${ llvm-mingw-toolchain-name }.tar.xz
    echo "${ llvm-mingw-toolchain-name }/bin" >> $GITHUB_PATH

# ...
```

riverar commented:
Looks good to me, just a nit on the caching that should be easy to resolve. Thanks for your patience while I was out!

mati865 commented:
Thanks, I'm out until Sunday/Morning. I'll address feedback then.

mati865 commented:
Sorry, was busier than anticipated.
Rebased branch (because of CI errors) and added cache:
![obraz](https://user-images.githubusercontent.com/1174646/182976280-e8d75c7a-267b-4640-89bd-ac410d4ab23c.png)

riverar commented:
Opened https://github.com/microsoft/windows-rs/pull/1949 to address build failures here.

riverar reviewed (changes requested):
Rebase once more (fix went in to master), should build cleanly now.

mati865 commented:
@riverar should be ready.

riverar commented:
@mati865 🥳🎉
